enclave/core/sgx/once.c

```diff
@@ -5,34 +5,74 @@
 
+    oe_once_t status = *once;
     /* Double checked locking (DCLP). */
-    oe_once_t o = *once;
-
     /* DCLP Acquire barrier. */
     OE_ATOMIC_MEMORY_BARRIER_ACQUIRE();
-    if (o == 0)
-    {
-        static oe_spinlock_t _lock = OE_SPINLOCK_INITIALIZER;
 
-        oe_spin_lock(&_lock);
-
-        if (*once == 0)
+    /*
+      Use an atomic-acquire load operation to check if the function has not been
+      invoked. If the function has already been invoked, there is nothing to do.
+      If the function is being invoked, then this thread must wait for the
+      function invocation to complete. Otherwise, this thread can try to take
+      ownership of invoking the function
+    */
+    if (status != FUNC_INVOKED)
+    {
+        /*
+          Multiple threads could reach here simultaneously after checking
+          whether the function has been invoked or not. Only one of them must
+          now invoke the function and others must wait for the function
+          invocation to complete.
+          To determine who gets to invoke the function, each thread atomically
+          compares the value of once to FUNC_NOT_INVOKED and if equal, sets the
+          value to FUNC_BEING_INVOKED to signal to other threads that the
+          function is being invoked. The return value of
+          __sync_val_compare_and_swap determines which thread takes ownership.
+          If __sync_val_compare_and_swap returns FUNC_NOT_INVOKED, then that
+          means this thread successfully set once to FUNC_BEING_INVOKED, and can
+          now safely call func. If __sync_val_compare_and_swap returns
+          FUNC_BEING_INVOKED, this means another thread's
+          __sync_val_compare_and_swap succeeded first and therefore this thread
+          now has to wait for the other thread to complete (ie wait for once to
+          become FUNC_INVOKED. If __sync_val_compare_and_swap returns
+          FUNC_INVOKED, that means another thread has already called the
+          function and marked the once as complete. This thread can safely
+          proceed.
+        */
+        oe_once_t retval = __sync_val_compare_and_swap(
+            once, FUNC_NOT_INVOKED, FUNC_BEING_INVOKED);
+        if (retval == FUNC_NOT_INVOKED)
         {
             if (func)
                 func();
 
-            /* DCLP Release barrier. */
             OE_ATOMIC_MEMORY_BARRIER_RELEASE();
-            *once = 1;
+            *once = FUNC_INVOKED;
+        }
+        else if (retval == FUNC_BEING_INVOKED)
+        {
+            /*
+              Another thread is invoking the function. Wait for that thread to
+              finish the invocation and mark the once variable to FUNC_INVOKED.
+            */
+            while (__sync_val_compare_and_swap(
+                       once, FUNC_BEING_INVOKED, FUNC_BEING_INVOKED) !=
+                   FUNC_BEING_INVOKED)
+            {
+                // Relinquish CPU
+                asm volatile("pause");
+            }
```

Comment:
@gupta-ak 
Thanks for catching this; it gave me an opportunity to re-scrutinize this.

The thread that sees _retval == FUNC_BEING_INVOKED_, infers that another thread is calling func and therefore decides to wait for *once to become FUNC_INVOKED (i.e for the invocation to complete).

It also needs to make sure it sees the side effects of the func too and not just the latest value of *once.
So it would need an _acquire barrier_ on line 75 just before exiting the scope.

What do you think?

Reply:
I believe GCC's CAS already has an `acquire/release` barrier in the implementation. We would need to just check the disassembly.

Reply:
GCC has deprecated the __sync_* line of intrinsic functions and replaced them with the __atomic_* line of functions.  The __atomic_* functions line up closely with the C++11 standards and allow you to explicitly declare the acquire/release barrier operations to use with each atomic operation that you call.

Reply:
As I mentioned 5 days back for the replacement to the _sync legacy function -
> @johnkord - Use the modern version i.e. __atomic_compare_exchange_n or __atomic_compare_exchange. Here is the reference. https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html
> Trust this helps.

Reply:
@EMumau Great suggestion. I have taken a stab at using the __atomic functions.
Can you do a quick review and also see if the comments look good as well?

@gupta-ak Can you also do a quick review? Let us wrap this up ASAP.